src/java.base/share/classes/java/lang/Float16.java

```diff
@@ -859,36 +873,188 @@ public static Float16 sqrt(Float16 radicand) {
-        // exponents can be held in a single double with precision to
-        // spare. Therefore, all the hard rounding cases should be
-        // covered, but some more analysis is needed to verify the
-        // correctness of that approach.
-        //
-        // Case analysis is needed with c is large compared to a*b.
-        //
-        // The number of significand
-        // bits in double, 53, is greater than the, maximum difference
-        // in exponent values between bit positions of minimum and
-        // maximum magnitude for Float16. Therefore, performing a*b+c
-        // in double and then doing a single rounding of that value to
-        // Float16 will implement this operation.
-
-        return valueOf( (a.doubleValue() * b.doubleValue()) +  c.doubleValue());
+        /*
+         * The double format has sufficient precision that a Float16
+         * fma can be computed by doing the arithmetic in double, with
+         * one rounding error for the sum, and then a second rounding
+         * error to round the product-sum to Float16. In pseudocode,
+         * this method is equivalent to the following code, assuming
+         * casting was defined between Float16 and double:
+         *
+         * double product = (double)a * (double)b;  // Always exact
+         * double productSum = product + (double)c;
+         * return (Float16)produdctSum;
+         *
+         * (Note that a similar relationship does *not* hold between
+         * the double format and computing a float fma.)
+         *
+         * Below is a sketch of the proof that simple double
+         * arithmetic can be used to implement a correctly rounded
+         * Float16 fma.
+         *
+         * ----------------------
+         *
+         * As preliminaries, the handling of NaN and Infinity
+         * arguments falls out as a consequence of general operation
+         * of non-finite values by double * and +. Any NaN argument to
+         * fma will lead to a NaN result, infinities will propagate or
+         * get turned into NaN as appropriate, etc.
+         *
+         * One or more zero arguments are also handled correctly,
+         * including the propagation of the sign of zero if all three
+         * arguments are zero.
+         *
+         * The double format has 53 logical bits of precision and its
+         * exponent range goes from -1022 to 1023. The Float16 format
+         * has 11 bits of logical precision and its exponent range
+         * goes from -14 to 15. Therefore, the bit positions
+         * representable in the Float16 format range from the
+         * subnormal 2^(-24), MIN_VALUE, to 2^15, the leading bit
+         * position of MAX_VALUE.
+         *
+         * Consequently, a double can hold the exact sum of any two
+         * Float16 values as the maximum difference in exponents of
+         * Float16 values less than the precision width of double.
```

Comment:
This is correct, but doesn't seem to be relevant for what follows. We don't have a sum between two `Float16` values here.